impersonation-filter/src/main/scala/org/openrepose/filters/impersonation/ImpersonationFilter.scala

```diff
@@ -0,0 +1,363 @@
+          getUser flatMap { userName =>
+            getImpersonationToken(getAdminToken, userName)
+          }
+
+        processingResult match {
+          case Success(_) =>
+            logger.trace("Response is valid")
+            Pass
+          case Failure(e: UserNotFoundException) =>
+            logger.trace(s"User not found: ${e.message}")
+            Reject(HttpServletResponse.SC_UNAUTHORIZED, Some(e.getMessage))
+          case Failure(e: MissingAuthTokenException) =>
+            logger.trace(s"Unauthorized: ${e.message}")
+            Reject(HttpServletResponse.SC_UNAUTHORIZED, Some(e.getMessage))
+          case Failure(e: IdentityCommunicationException) =>
+            logger.trace(s"Bad gateway: ${e.message}")
+            Reject(HttpServletResponse.SC_BAD_GATEWAY, Some(e.getMessage))
+          case Failure(e: OverLimitException) =>
+            logger.trace(s"Rate limited: ${e.message}")
+            response.addHeader(HttpHeaders.RETRY_AFTER, e.retryAfter)
+            Reject(HttpServletResponse.SC_SERVICE_UNAVAILABLE, Some(e.getMessage))
+          //case Failure(e) if e.getCause.isInstanceOf[AkkaServiceClientException] && e.getCause.getCause.isInstanceOf[TimeoutException] =>
+          //  Reject(HttpServletResponse.SC_GATEWAY_TIMEOUT, Some(s"Call timed out: ${e.getMessage}"))
+          case Failure(e) =>
+            logger.trace(s"Something else: ${e.getMessage}")
+            Reject(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, Some(e.getMessage))
+        }
+      }
+
+      filterResult match {
+        case Pass =>
+          logger.trace("Processing completed, passing to next filter or service")
+          chain.doFilter(request, response)
+        case Reject(statusCode, message) =>
+          Option(config.getDelegating) match {
+            case Some(delegating) =>
+              logger.debug(s"Delegating with status $statusCode caused by: ${message.getOrElse("unspecified")}")
+              val delegationHeaders = buildDelegationHeaders(statusCode,
+                "rackspace-impersonation",
+                message.getOrElse("Failure in the Rackspace impersonation filter").replace("\n", " "),
+                delegating.getQuality)
+
+              delegationHeaders foreach { case (key, values) =>
+                values foreach { value =>
+                  request.addHeader(key, value)
+                }
+              }
+              chain.doFilter(request, response)
+            case None =>
+              logger.debug(s"Rejecting with status $statusCode")
+
+              message match {
+                case Some(m) =>
+                  logger.debug(s"Rejection message: $m")
+                  response.sendError(statusCode, m)
+                case None => response.sendError(statusCode)
+              }
+          }
+      }
+    }
+
+    def updateTokenHeader(impersonationToken: ImpersonationToken): Unit = {
+      if(request.getHeader(CommonHttpHeader.AUTH_TOKEN.toString) != null)
```

Comment:
Don't check for null in scala.

instead:
```scala

Option(request.getHeader(CommonHttpHeader.AUTH_TOKEN.toString)) map { token =>
        request.replaceHeader(CommonHttpHeader.AUTH_TOKEN.toString, impersonationToken.token)
} getOrElse {
        request.addHeader(CommonHttpHeader.AUTH_TOKEN.toString, impersonationToken.token) 
}
```

Or since this is entirely about side effects:
```scala
if(Option(request.getHeader(CommonHttpHeader.AUTH_TOKEN.toString)).exists) {
        request.replaceHeader(CommonHttpHeader.AUTH_TOKEN.toString, impersonationToken.token)
} else {
        request.addHeader(CommonHttpHeader.AUTH_TOKEN.toString, impersonationToken.token) 
}
```